File: javascript/solve_for_x.js
// solve for x
// given: 2x-4=0

const solve_x = eq => {
  eq = eq.replace(/(\-|\+)(?=[a-z])/, c=>"1"+c).replace(/^[a-z]/, c=>"1"+c)
  const sides = eq.split("=")

  const regex_int = /(\+|\-)?\d+(?![a-z])/g
  
  const regex_strn = /(\+|\-)?\d+(?=[a-z])/g

  let ints = sides.map(side => side.match(regex_int)).filter(s=>s).map(x => {
    return x.map(Number).reduce((s,v)=>s+v,0)
  })
  ints = ints.length > 1 ? ints.map((v,i) => i === 0 ? -v : v).reduce((s,v)=>s+v) : ints.reduce((s,v)=>s+v)
                                                                         
  let strn = sides.map(side => side.match(regex_strn)).filter(s=>s).map(x => {
    return x.map(Number).reduce((s,v)=>s+v,0)
  })
  strn = strn.length > 1 ? strn.map((v,i) => i !== 0 ? -v : v).reduce((s,v)=>s+v) : strn.reduce((s,v)=>s+v)
  
  return ints/strn
}

// test the solution
console.log(solve_x("2x-4=0"))

// more examples
console.log(solve_x("3x-4+5=2x-5"))
console.log(solve_x("-4x-4+5=2x"))
console.log(solve_x("-1x-5=2x"))

// future self
// make this work
// => x-2=0
console.log(solve_x("x-2=0"))
// currently needs to be like this to work
// => 1x-2=0

File: python/solve_for_x.py
# solve for x
# 2x-4=0
import re
from functools import reduce

def solve_x(txt):
    # regex_char = re.compile("([-|\+])(?=[a-z])")
    # print(regex_char.findall(txt))
    # txt = re.sub(regex_char, lambda c: "1"+c, txt)
    print("txt:", txt)
    sides = txt.split("=")
    # return list(map(lambda side: side[::-1], sides))
    
    regex_strn = re.compile(r'([-|\+]?[0-9]+)(?=[a-z])')
    regex_int = re.compile(r'([-|\+]?[0-9]+)(?![a-z])')
    
    ints = list(filter(lambda a: a, list(map(lambda side: regex_int.findall(side), sides))))
    ints = [reduce(lambda a,b:a+b, lst, 0) for lst in [[int(x) for x in lst] for lst in ints]]
    new_ints = [] 
    for i, x in enumerate(ints):
        if(len(ints)) > 1:
            new_ints.append(x if i != 0 else -x)

    strn = list(filter(lambda a: a, list(map(lambda side: regex_strn.findall(side), sides))))
    strn = [reduce(lambda a,b:a+b, lst, 0) for lst in [[int(x) for x in lst] for lst in strn]]
    new_strn = [] 
    for i, x in enumerate(strn):
        if(len(ints)) > 1:
            new_strn.append(-x if i != 0 else x)
            
    ints = reduce(lambda a,b:a+b, new_ints, 0)
    strn = reduce(lambda a,b:a+b, new_strn, 0)
    
    print("strn:", new_strn)
    print("ints:", new_ints)
    print("strn:", strn)
    print("ints:", ints)
    return ints/strn
    
print(solve_x("1x-2=0"))
print("---")
print(solve_x("2x-4=2"))